guide: Comparing Experiments (#2862)
iterative/dvc.org: content/docs/user-guide/experiment-management/comparing-experiments.md

# Comparing Experiments

You'll need a way to compare experiment results and track their corresponding
parameters. DVC provides commands to list, tabulate and contrast them. Let's see
how they can help you streamline the experimentation process.

## List experiments in the workspace

You can get a list of existing experiments in the repository with
`dvc exp list`. Without any options, this command lists the experiments based on
the latest commit of the current branch (Git `HEAD`).

```dvc
$ dvc exp list
refs/tags/baseline-experiment:
        cnn-64
        cnn-128
```

If you want to list all the experiments in the repository regardless of their
parent commit, use the `--all` flag.

```dvc
$ dvc exp list --all
refs/tags/baseline-experiment:
        cnn-64
        cnn-128
main:
        exp-93150
```

## List experiments in another Git remote

As we discuss in [Sharing Experiments], you can use `dvc exp push` to upload
experiments to Git remotes. `dvc exp list` can be used to list the experiments
in a Git remote.

```dvc
$ dvc exp list origin
refs/tags/baseline-experiment:
        cnn-32
        cnn-64
```

> This command lists the experiments originated from `HEAD`. You can add any other

Reviewer: @iesahin another question here... do you know how `HEAD` is determined for a remote to list experiments from? Is it the local repo's HEAD? Is it the default branch HEAD always? We should probably clarify (let's add it to https://github.com/iterative/dvc.org/issues/2995).

Reviewer: p.s. actually we could add it in https://github.com/iterative/dvc.org/pull/2908 as it's more about sharing than about comparing.

Reviewer: Moved to https://github.com/iterative/dvc.org/pull/2908#pullrequestreview-798148999.

layout. When you want to use the output in other commands, `dvc exp diff` can
output in JSON with `--json` flag.

```dvc
$ dvc exp diff exp-25a26 cnn-64 --json
```

```json
{
  "metrics": {
    "metrics.json": {
      "loss": {
        "old": 0.2386719286441803,
        "new": 0.2338544875383377,
        "diff": -0.00481744110584259
      },
      "acc": {
        "old": 0.9150999784469604,
        "new": 0.9153000116348267,
        "diff": 0.00020003318786621094
      }
    }
  },
  "params": {
    "params.yaml": {
      "model.conv_units": {
        "old": 256,
        "new": 64,
        "diff": -192
      }
    }
  }
}
```

The output is a JSON dictionary with two keys, `metrics` and `params`, which
have dictionaries as values. `metrics` and `params` dictionaries has keys for
each of the metrics or params file, and for each file metrics and parameters are
listed as keys.

As an example, we can get only a specific metric [jq]:

```dvc
$ dvc exp diff exp-25a26 cnn-64 --json | jq '.metrics."metrics.json".acc'
```

```json
{
  "old": 0.9150999784469604,
  "new": 0.9153000116348267,
  "diff": 0.00020003318786621094
}
```

### Get the comparison in Markdown table format

> table to embed in the reports directly.

Reviewer: Re-add line that makes the link checker fail
```suggestion
`dvc exp diff` has also an option to output a [Github-Flavored Markdown][gfm]
table to embed in the reports directly.
```

Path         Param             Value    Change
params.yaml  model.conv_units  256      240
params.yaml  train.epochs      10       0
```

## Compare an experiment with the workspace

When you want to compare two experiments, either the baseline experiment in a
commit, branch, tag or an attached experiment with ID, you can supply their
names to `dvc exp diff`.

```
$ dvc exp diff exp-25a26 cnn-64
Path          Metric    Value    Change
metrics.json  acc       0.9153   0.00020003
metrics.json  loss      0.23385  -0.0048174

Path         Param             Value    Change
params.yaml  model.conv_units  64       -192
```

## Customize the comparison output

You can hide the file path of metrics and parameter files the `--no-path` option
-- produces cleaner output for single params/metrics files, for example. You can
also customize the amount of significant digits shown for numeric values with
the `--precision` option (5 by default).

```dvc
$ dvc exp diff exp-25a26 cnn-64 --no-path --precision 2
Metric    Value    Change
acc       0.92     0.0002
loss      0.23     -0.0048

Param             Value    Change
model.conv_units  64       -192
```

### Get the comparison in JSON

> Parsing the `dvc exp diff` output may not be feasible due to the custom table
> layout. When you want to use the output in other commands, `dvc exp diff` can

Reviewer: 💅 "custom" sounds like user customized something but we're referring to the output of `exp diff` in general.
```suggestion
Parsing the output of `dvc exp diff` may not be easy when
you want to use it in other commands. `dvc exp diff` can
```

output in JSON with `--json` flag.

```dvc
$ dvc exp diff exp-25a26 cnn-64 --json
```

```json
{
  "metrics": {
    "metrics.json": {
      "loss": {
        "old": 0.2386719286441803,
        "new": 0.2338544875383377,
        "diff": -0.00481744110584259
      },
      "acc": {
        "old": 0.9150999784469604,
        "new": 0.9153000116348267,
        "diff": 0.00020003318786621094
      }
    }
  },
  "params": {
    "params.yaml": {
      "model.conv_units": {
        "old": 256,
        "new": 64,
        "diff": -192
      }
    }
  }
}
```

The output is a JSON dictionary with two keys, `metrics` and `params`, which
have dictionaries as values. `metrics` and `params` dictionaries has keys for
each of the metrics or params file, and for each file metrics and parameters are
listed as keys.

> As an example, we can get only a specific metric [jq]:
> 
> ```dvc
> $ dvc exp diff exp-25a26 cnn-64 --json | jq '.metrics."metrics.json".acc'
> ```
> 
> ```json
> {

Reviewer: CE
```suggestion
As an example, we can get only a specific metric with [jq]:

```dvc
$ dvc exp diff exp-25a26 cnn-64 --json | jq '.metrics."metrics.json".acc'
{
```

  "old": 0.9150999784469604,
  "new": 0.9153000116348267,
  "diff": 0.00020003318786621094
}
```

### Get the comparison in Markdown table format

> `dvc exp diff` has also an option to output a [Github-Flavored Markdown][gfm]

Reviewer: To test out https://github.com/iterative/dvc.org/pull/2862#issuecomment-954083610 ⌛ 
```suggestion

```

> `dvc exp diff` has also an option to output a [Github-Flavored Markdown][gfm]

Reviewer: Yep, this is the line that makes the link checker fail. I removed it and it's passing now. Will re-add it though (we can merge as-is)

cnn-64,6d13f33,branch_commit,2021-09-09T13:06:05,,0.2338544875383377,0.9153000116348267,10,64
cnn-128,69503c6,branch_commit,2021-09-09T12:53:51,,0.2324332743883133,0.9160000085830688,10,128
cnn-96,49779aa,branch_base,2021-09-09T12:53:51,,0.2309877723455429,0.916100025177002,10,96
```

You can supply this output to other commands as well. For example with [csvkit],
you can get a summary statistics about the experiments.

```dvc
$ dvc exp show --csv | csvstat
...
7. "acc"

        Type of data:          Number
        Contains null values:  False
        Unique values:         5
        Smallest value:        0.9127
        Largest value:         0.9167
        Sum:                   5.4895
        Mean:                  0.914917
        Median:                0.91565
        StDev:                 0.001774
        Most common values:    0.9127 (2x)
                               0.9167 (1x)
                               0.9153 (1x)
                               0.9161 (1x)
                               0.916 (1x)
...
```

[csvkit]: https://csvkit.readthedocs.io/en/latest/

## Compare specific experiments

> In addition to showing a summary table of experiments, DVC provides a command to
> compare pairwise experiments. `dvc exp diff` is used to compare the experiments
> by the change in their metrics and params.

Reviewer: > If we reorder the sections

BTW I thought we were doing that but I guess not (for now) so just this:

```suggestion
In addition to showing a summary table of experiments, DVC provides the
`dvc exp diff` command to compare pairs of experiments by the difference
in their metrics and params.
```

> by the change in their metrics and params.

Reviewer: > committing along with a few uncontested copy edits ⌛

```dvc
$ dvc exp diff
Path          Metric    Value    Change
metrics.json  acc       0.9151   0.0024
metrics.json  loss      0.23867  0.0020977

Path         Param             Value    Change
params.yaml  model.conv_units  256      240
```

Without experiment names, `dvc exp diff` shows the change between the last
experiment and the last commit. So if the command doesn't print an output, there
might be no experiment since the previous commit or it didn't produce changes in
results. If you want to see all the parameters and metrics regardless of whether
they have changed, you can use `--all` flag.

```dvc
$ dvc exp diff --all
Path          Metric    Value    Change
metrics.json  acc       0.9151   0.0024
metrics.json  loss      0.23867  0.0020977

Path         Param             Value    Change
params.yaml  model.conv_units  256      240
params.yaml  train.epochs      10       0
```

## Compare an experiment with the workspace

When you want to compare two experiments, either the baseline experiment in a
commit, branch, tag or an attached experiment with ID, you can supply their

> names to `dvc exp diff`.

Reviewer: If this is about comparing vs. workspace, why are we supplying 2 experiments to `exp diff` below? The only way to reliably compare the workspace (to `HEAD` aka baseline) is to use plan `exp diff` without arguments, no?

> names to `dvc exp diff`.

Reviewer 2: Agreed with @jorgeorpinel. The headers of these two sections seem reversed from the examples provided. "Compare specific experiments" shows a comparison to the workspace, and "Compare an experiment with the workspace" shows comparing two experiment revisions (not the workspace).

```

> $ dvc exp diff exp-25a26 cnn-64

Reviewer: Let's just use another existing one for consistency for now? It's hypothetical anyway
```suggestion
$ dvc exp diff cnn-128 cnn-64
```

Metric    Value    Change
acc       0.92     0.0002
loss      0.23     -0.0048

Param             Value    Change
model.conv_units  64       -192
```

### Get a JSON list of changes to use in scripts

Parsing the `dvc exp diff` output may not be feasible due to the custom
structure. When you want to use the output in other commands, `dvc exp diff` can
output in JSON with `--json` flag.

```dvc
$ dvc exp diff exp-25a26 cnn-64 --json | jq
```

```json
{
  "metrics": {
    "metrics.json": {
      "loss": {
        "old": 0.2386719286441803,
        "new": 0.2338544875383377,
        "diff": -0.00481744110584259
      },
      "acc": {
        "old": 0.9150999784469604,
        "new": 0.9153000116348267,
        "diff": 0.00020003318786621094
      }
    }
  },
  "params": {
    "params.yaml": {
      "model.conv_units": {
        "old": 256,
        "new": 64,
        "diff": -192
      }
    }
  }
}
```

The output is a JSON dictionary with two keys, `metrics` and `params`, which
have dictionaries as values. `metrics` and `params` dictionaries has keys for
each of the metrics or params file, and for each file metrics and parameters are

> listed as keys.

Reviewer: Thinking about it again, I don't think we need to document these output schemas that formally. Comprehensive examples (we got those already) should be enough.

the columns by params or metrics by the option `--sort-by` and `--sort-order`.
`--sort-by` takes a metric or parameter name as defined in `dvc.yaml` and
`--sort-order` takes either `asc` or `desc`.

```dvc
$ dvc exp show --sort-by auc --sort-order desc
````

```dvctable
┏━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━┳━━━━━━━━━━━━━━━━━━┓
┃ neutral:**Experiment**            ┃    metric:**acc** ┃ param:**model.conv_units** ┃
┡━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━╇━━━━━━━━━━━━━━━━━━┩
│ workspace             │ 0.9127 │ 16               │
│ baseline-experiment   │ 0.9127 │ 16               │
│ ├── ca55d7d [cnn-32]  │ 0.9167 │ 32               │
│ ├── 6d13f33 [cnn-64]  │ 0.9153 │ 64               │
│ ├── 49779aa [cnn-96]  │ 0.9161 │ 96               │
│ └── 69503c6 [cnn-128] │  0.916 │ 128              │
└───────────────────────┴────────┴──────────────────┘
```

## Get experiments table in JSON

There may be times when a machine readable format for the experiments is
required. In its default settings `dvc exp show` lists the experiments in a
visually appealing way that may not be suitable to parse in scripts. To get a
list of experiments with their params and metrics, you can use `--show-json` or

> `--show-csv` flags.

Reviewer: It'll be `--json` and `--cvs` soon (as hinted in https://github.com/iterative/dvc.org/pull/2862#issuecomment-932408692) ☝️

# Comparing Experiments

After running the experiments, it's desirable to compare them by their
parameters and the corresponding results. DVC provides three commands to list,
tabulate and compare the experiments. In this section, we discuss various use
cases and options to streamline experimentation.

## List experiments in the workspace

After running the experiments, you can get a list of them by `dvc exp list`.
Without any options, `dvc exp list` lists the experiments after the most recent
commit.

```dvc
$ dvc exp list
refs/tags/baseline-experiment:
        cnn-32
        cnn-64
        cnn-96
        cnn-128
```

If you want to list all the experiments in the repository independent of the
commits, you can use `--all` flag.

```dvc
$ dvc exp list --all
refs/tags/baseline-experiment:
        cnn-64
        cnn-128
        cnn-32
        cnn-96
b99a33e:
        exp-93150
```

## List experiments in another Git remote

As we discuss in [Sharing Experiments], you can use `dvc exp push` to upload
experiments to Git remotes. `dvc exp list` can be used to list the experiments

> in a Git remote.

Reviewer: p.s. this is going away in https://github.com/iterative/dvc.org/pull/2908 anyway 🤷. Resolving

# Comparing Experiments

After running the experiments, it's desirable to compare them by their
parameters and the corresponding results. DVC provides three commands to list,
tabulate and compare the experiments. In this section, we discuss various use
cases and options to streamline experimentation.

## List experiments in the workspace

After running the experiments, you can get a list of them by `dvc exp list`.
Without any options, `dvc exp list` lists the experiments after the most recent
commit.

```dvc
$ dvc exp list

> refs/tags/baseline-experiment:

Reviewer: WDYT here @dberenbaum ? Rel. the "baseline" concept/term in general. Thanks

> refs/tags/baseline-experiment:

Reviewer 2: I don't have strong opinion. Baseline makes sense to me, and I can see the argument for and against calling it an experiment. Calling it a baseline experiment seems fine. Having `baseline-experiment` as the tag feels a bit long and artificial, but again it's not a strong opinion. Seems like we have enough other areas where we can improve our experiments documentation that it's not worth spending too much time debating this. Hopefully we will get more crisp in how to explain experiments over time, seeing what works and what causes confusion.
